Add crude autocomplete and history

geowarin opened the pull request:
This adds a very rudimentary history that you can navigate using the UP and DOWN arrows.
This is allowed only when the caret is at position 0.

I did not prevent UP and DOWN from going to the beginning of the line_edit and the end, respectively.
I like it like this but, we can easily change this I think.

History size is unlimited, I don't think this is going to be a problem.

Using the TAB key triggers autocompletion. The matching function is very simple as well, we could improve it, no doubt.

I added focus_next and focus_previous on the line_edit that point to itself to prevent focus loss... We could easily regrab focus instead.
We need a call deferred function to set the caret at the end anyway.

I'm not sure why there are more diff in the default_ui.tscn, this is not my doing :smile: 
I can revert it if convenient.

Anyway, love the plugin!
Waiting for your feedback!

NB: I'm relatively new to gdscript so feel free to point out my mistakes :wink:

geowarin committed:
Add crude autocomplete and history

Kubulambula commented:
Thank's for the PR! :heart:

Don't worry about the `default_ui.tscn` diff. There was an issue with older Godot 4 versions, that created a lot of diffs in the same file without any effect on each save. It's fixed now, but I did not change the file since then, so just commit it and it should be good in the future.


Anyway I like it overall but, there are some things, that should be addressed first.

1. ---
I think that both autocomplete and history should be handled inside `GDShellUIHandler` as it is just a user side function, that would not make sense if the GDShell is used via buttons for example. Also `GDShellMain` is bloated enough already, so having the code elsewhere would be better.

It would be great if the `GDShellUIHandler` would give the user some interface to work with history and autocompletion, like `history_get_next()` function that the user can call when UP is pressed and work with the result.
The history could be tracked in the background even if the user does not use it via `GDShellUIHandler.submit_input()`.

Maybe give the user a `History` and `Autocompletion` objects, that would handle all of this?


2. ---
Consider this entered sequence:
```
echo a
echo b
echo c
```
Then you would press UP and end up on `echo b`. Then you would clear the `LineEdit` and then press UP again. You are now on `echo a`. 
I'm not entirely sure how to fix it right now, as the UI can do whatever it wants and you won't know that the input was cleared. It could be resolved with the new way of handling history and the UI would take responsibility to notify the history that next time it should start over?


3. ---
Consider the same entered sequence:
```
echo a
echo b
echo c
```
Then press 3x UP (end up on `echo a`) and then press 2x DOWN. You should end up on `echo c` again, but you get stuck on `echo b`, because DOWN is shortcut to move the cursor to the end and UP is shortcut for moving the cursor to the beginning of the `LineEdit`

geowarin committed:
address feedback

Pressing up/down leaves the caret at position 0
Pressing any key other than up/down/tab clears the position in history
Code moved to gdshell_ui_handler

geowarin commented:
Thanks for the feedback, I updated the code a little bit.
I agree that it looks a bit messy, introducing history and completion classes would help a little.

I'd love to move the handler `_on_input_line_edit_gui_input` but I'm not sure if there's a better place for it.

geowarin commented:
What would be awesome is a way for commands to provide completions for their arguments.

Let say for instance that I have a command `load_level` that takes a level_name as a parameter.
I'd like this command to provide autocomplete for the level names.

That is probably a bit more tricky as the autocomplete function would need to parse the current buffer to see if it begins with a known command and then delegate autocompletion to this command.

Is that something you would be interested in? How would you tackle this?

We could also persist the history between sessions. That would be practical I think.

Kubulambula commented:
Sorry that I couldn't get to you sooner.

I like how it interfaces the `GDShellUIHandler` now, but I still have something to say.
When you use the history a few times and then decide that instead of finding the command in the endless history, you write the command again yourself and do `Ctrl+A` and then `Delete` to clear the text. Because the cursor is focused to the beginning, the history is not restarted.

I suggest moving the cursor to the back on every history scroll instead to the front of the `LineEdit` and delete the condition that checks if you are at the beggining of the `LineEdit` so that you can engage the history at any time.

Kubulambula commented on addons/gdshell/ui/default_ui/default_ui.gd:
```diff
@@ -123,3 +123,29 @@ func _on_visibility_changed() -> void:
+		if event.keycode == KEY_TAB:
+			input_line_edit.text = autocomplete(text_before_caret())
+			set_line_edit_caret_to_end.call_deferred()
+		elif input_line_edit.caret_column == 0:
+			if event.keycode == KEY_UP:
+				input_line_edit.text = history_get_next()
+				set_line_edit_caret_to_beginning.call_deferred()
+			elif event.keycode == KEY_DOWN:
+				input_line_edit.text = history_get_previous()
+				set_line_edit_caret_to_beginning.call_deferred()
```
These are the changes mentioned in the previous comment. With this, I think that the history good
```suggestion
	elif event.keycode == KEY_UP:
		input_line_edit.text = history_get_next()
		set_line_edit_caret_to_end.call_deferred()
	elif event.keycode == KEY_DOWN:
		input_line_edit.text = history_get_previous()
		set_line_edit_caret_to_end.call_deferred()
```

geowarin committed:
Update addons/gdshell/ui/default_ui/default_ui.gd

Co-authored-by: Jakub Janšta <kubulambula@gmail.com>

geowarin committed:
Fix indent

Kubulambula commented:
> What would be awesome is a way for commands to provide completions for their arguments.

That would be really awesome indeed! 

The `autocomplete` function would take what's already in the `LineEdit` and possibly use `GDShellCommandParser`. Then it would try to instance the command beforehand in the limited state that it is in, when it gets loaded by `GDShellCommandDB` to get the autoaliases, delegate the autocomplete to that instance and when it is complete, free it.

However the autocomplete functions inside commands would become really complicated. Imagine that you have 10 possible arguments. You would have to determine what argument are you autocompleting and then do the search.

I think that it is doable

---

> We could also persist the history between sessions. That would be practical I think.

The history could be stored in the user:// directory so that it could be used even in exported builds. There could also be be stored some global GDShell "ENV" variables if they come to existence.

I would just create a `gdshell` folder right next to `logs` and `shader_cache` and have some file named `history` in there, that would be read on startup and updated on exit.

I would make this in a new PR tho.

Kubulambula commented:
I think that for now it's more than enough!

Thanks! ❤️
